### PR TITLE
Made strmakeupper() and strmakelower() be Unicode aware

### PR DESCRIPTION
**Pull request by npwoods**

### CODE REVIEW

File: src/lib/util/corestr.cpp [hunk 1]
```diff
@@ -206,12 +208,49 @@ std::string &strtrimrightspace(std::string& str)
 	return internal_strtrimspace(str, true);
 }
 
-std::string &strmakeupper(std::string& str)
+static std::string &strtransform_uchar(std::string& str, char32_t(*callback)(char32_t ch))
```
Comment (cuavas) [inline src/lib/util/corestr.cpp]: This can't be flattened if it's not templated on the mutator type.

File: src/lib/util/corestr.cpp [hunk 2]
```diff
@@ -206,12 +208,49 @@ std::string &strtrimrightspace(std::string& str)
...
 {
-	std::transform(str.begin(), str.end(), str.begin(), ::toupper);
+	size_t i = 0;
+	while(i < str.size())
```
Comment (cuavas) [inline src/lib/util/corestr.cpp]: Inconsistent spacing.

File: src/lib/util/corestr.cpp [hunk 3]
```diff
@@ -206,12 +208,49 @@ std::string &strtrimrightspace(std::string& str)
...
+		char32_t old_character;
+		int rc = uchar_from_utf8(&old_character, &str[i], str.size() - i);
+
+		if (rc > 0)
+		{
+			char32_t new_character = callback(old_character);
+			if (old_character != new_character)
+			{
+				// if the character has changed, replace it
+				std::string new_character_string = utf8_from_uchar(new_character);
```
Comment (cuavas) [inline src/lib/util/corestr.cpp]: This is horribly inefficient - you're doing moves all over the place.  You're better off operating on ranges.  Start with an empty transformed string, when you get a character that needs transforming, copy the entire range from start to the changing character, then append the changed character, etc.  If you want to get fancy, you can also copy ranges only at points where the size of the UTF-8 sequence changes.

**Comment by npwoods** [comment]
While the proposed technique will also work, isn't it in practice less efficient when one takes into account the nature of changing upper/lower case?  Specifically, I am assuming that for the most part, changing a character from lower to upper case or lower to upper case will not change the UTF-8 byte length,

Examples:
- `A` ==> `a`:  0x45 ==>0x65
- `Ā` ==> `ā`:  0xC4 0x80 ==> 0xC4 0x81

I am making a leap of faith and assuming that `std::string::replace` is particularly smart about dropping bytes into existing strings when it is replacing N characters with N characters, and obviously it will be slower but do the right thing if this isn't the case.

The approach you proposed will also work, but always involve copying the string.

**Comment by cuavas** [comment]
Well, even if we assume that the replacement is going to be efficient, why even use a `std::string` in the loop there?  It requires destruction each time it falls out of scope, you're adding at least one exception unwind landing point, and you're using a less-efficient version of `utf8_from_uchar`.  You know what the longest possible UTF-8 sequence is, you can use a `char` array declared outside the loop, the more efficient overload of `utf8_from_uchar` and use one of the other overloads of `std::string::replace` (e.g. [overload 4 on this page](http://en.cppreference.com/w/cpp/string/basic_string/replace)).

**Pull request closed without merging**